wireless_watcher/src/wireless_watcher/mikrotik_api.py

```diff
@@ -0,0 +1,216 @@
+#     from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+#  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+#  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+#  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+#  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+#  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+#  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+#  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+
+import binascii
+import md5
+import posix
+import time
+import socket
+import select
+import sys
+
+
+class RouterOSApi(object):
+    "Routeros api"
+    def __init__(self, hostname, port=8728):
+        self.sk = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.sk.connect((hostname, port))
+        self.currenttag = 0
+
+    def login(self, username="admin", pwd="", verbose=True):
+        for repl, attrs in self.talk(["/login", "=name=" + username, "=password=" + pwd], verbose):
+            if repl == '!trap':
+                raise ValueError("Failed to login, check username and password and try again.")
+            elif '=ret' in attrs.keys():
+                #for repl, attrs in self.talk(["/login"]):
+                chal = binascii.unhexlify(attrs['=ret'])
+                md = md5.new()
+                md.update('\x00')
+                md.update(pwd)
+                md.update(chal)
+                for repl2, attrs2 in self.talk(["/login", "=name=" + username,
+                                               "=response=00" + binascii.hexlify(md.digest())], verbose):
+                    if repl2 == '!trap':
+                        return False
+        return True
+
+    def talk(self, words, verbose=True):
+        if self.writeSentence(words, verbose) == 0:
+            return
+        r = []
+        while 1:
+            i = self.readSentence(verbose)
+            if len(i) == 0:
+                continue
+            reply = i[0]
+            attrs = {}
+            for w in i[1:]:
+                j = w.find('=', 1)
+                if (j == -1):
+                    attrs[w] = ''
+                else:
+                    attrs[w[:j]] = w[j+1:]
+            r.append((reply, attrs))
+            if reply == '!done':
+                return r
+
+    def writeSentence(self, words, verbose):
+        ret = 0
+        for w in words:
+            self.writeWord(w, verbose)
+            ret += 1
+        self.writeWord('', verbose)
+        return ret
+
+    def readSentence(self, verbose=True):
+        r = []
+        while 1:
```

Comment:
All of these could probably be written as generators. But it's fine as is.

Reply:
You seem to have a preference for generators. Can you expand on why they would be preferable here?

Reply:
They're just more pythonic. Which is why I "prefer" them but am not about to be a zealot about it.  Any time we're basically munching a data stream until it's empty, that's what a generator is for.  Being more Pythonic basically just means being more maintainable and universally legible.